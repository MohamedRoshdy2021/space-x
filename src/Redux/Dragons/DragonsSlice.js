import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

const url = "https://api.spacexdata.com/v4/dragons";
const initialState = {
  isLoading: false,
  data: [],
  error: "",
  joinedDragons: [],
};
const getData = (data) =>
  data.map((dragons) => ({
    id: dragons.id,
    name: dragons.name,
    description: dragons.description,
    image: dragons.flickr_images,
    reserved: false,
   
  }));
export const fetchDragonsData = createAsyncThunk(
  "dragons/fetchDragons",
  async () => {
    try {
      const response = await axios.get(url);
      const { data } = response;
      return getData(data);
    } catch (error) {
      throw new Error(error);
    }
  }
);
const dragonSlice = createSlice({
  name: "dragons",
  initialState,
  reducers: {
    reserveDragon: (state, action) => {
      const newState = state.data.map((dragon) => {
        if (dragon.id === action.payload) {
          return { ...dragon, reserved: true };
        }
        return dragon;
      });
      return {
        ...state,
        data: newState,
        joinedDragons: [...state.joinedDragons, action.payload],
      };
    },
    cancelReserveDragon: (state, action) => {
      const newState = state.data.map((dragon) => {
        if (dragon.id === action.payload) {
          return { ...dragon, reserved: false };
        }
        return dragon;
      });
      return {
        ...state,
        data: newState,
        joinedDragons: state.joinedDragons.filter(
          (dragonId) => dragonId !== action.payload
        ),
      };
    },
  },
  
  extraReducers: (builder) => {
    builder
        .addCase(fetchDragonsData.pending, (state) => ({
            ...state,
            isLoading: true,
        }))
        .addCase(fetchDragonsData.fulfilled, (state, action) => ({
            ...state,
            isLoading: false,
            data: action.payload,

        }))
        .addCase(fetchDragonsData.rejected, (state,action) => ({
            ...state,
            isLoading: false,
            error: action.error.message,
        }));
  }


});

export const { reserveDragon, cancelReserveDragon } = dragonSlice.actions;
export default dragonSlice.reducer;
