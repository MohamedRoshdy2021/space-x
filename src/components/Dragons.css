.dragons-container {
  padding: 10px 5%;
}

.dragon-left {
  display: flex;
  gap: 20px;
  margin-bottom: 40px;
}

.dragon-image {
  width: 400px;
}

.dragon-btn {
  padding: 10px;
  background-color: #1c7ed6;
  color: white;
  border: none;
  border-radius: 4px;
  font-size: 20px;
  margin-top: 20px;
}

.dragon-name,
.dragon-desc {
  color: #333;
}

.reserved-badge {
  padding: 1px 4px;
  border: 1px solid #22b8cf;
  background-color: #22b8cf;
  border-radius: 4px;
  font-size: 14px;
  margin-right: 10px;
  color: white;
}

.cancel {
  background-color: white;
  color: #777;
  border: 1px solid #777;
}
