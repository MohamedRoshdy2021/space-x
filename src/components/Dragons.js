function Dragons() {
  return (
    <div>
      <h1>This is the Dragons page</h1>
    </div>
  );
}
export default Dragons;
