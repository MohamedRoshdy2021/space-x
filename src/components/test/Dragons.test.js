import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Dragons from '../Dragons';

const mockStore = configureStore([thunk]);

test('matches snapshot for Dragons component', () => {
    const store = mockStore({
      dragons: {
        isLoading: false,
        data: [
          {
            id: 1,
            name: 'Dragon 1',
            description: 'Description 1',
            image: 'dragon1.jpg',
          },
          {
            id: 2,
            name: 'Dragon 2',
            description: 'Description 2',
            image: 'dragon2.jpg',
          },
        ],
        joinedDragons: [1],
      },
    });
  
    const { asFragment } = render(
      <Provider store={store}>
        <Dragons />
      </Provider>,
    );
  
    expect(asFragment()).toMatchSnapshot();
  });