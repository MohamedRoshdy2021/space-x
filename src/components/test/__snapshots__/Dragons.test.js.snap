// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot for Dragons component 1`] = `
<DocumentFragment>
  <div
    class="dragons-container"
  >
    <div
      class="dragon-left"
    >
      <img
        alt="Dragon 1"
        class="dragon-image"
        src="dragon1.jpg"
      />
      <div
        class="dragon-details"
      >
        <h2
          class="dragon-name"
        >
          Dragon 1
        </h2>
        <p
          class="dragon-desc"
        >
           
          <span
            class="reserved-badge"
          >
            Reserved
          </span>
          Description 1
        </p>
        <button
          class="dragon-btn cancel"
          data-testid="cancel-reservation-button-1"
          type="button"
        >
          Cancel Reservation
        </button>
      </div>
    </div>
    <div
      class="dragon-left"
    >
      <img
        alt="Dragon 2"
        class="dragon-image"
        src="dragon2.jpg"
      />
      <div
        class="dragon-details"
      >
        <h2
          class="dragon-name"
        >
          Dragon 2
        </h2>
        <p
          class="dragon-desc"
        >
          Description 2
        </p>
        <button
          class="dragon-btn"
          data-testid="reserve-button-2"
          type="button"
        >
          Reserve dragon
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot for Rockets component 1`] = `
<DocumentFragment>
  <div
    class="dragons-container"
  >
    <div
      class="dragon-left"
    >
      <img
        alt="Dragon 1"
        class="dragon-image"
        src="dragon1.jpg"
      />
      <div
        class="dragon-details"
      >
        <h2
          class="dragon-name"
        >
          Dragon 1
        </h2>
        <p
          class="dragon-desc"
        >
           
          <span
            class="reserved-badge"
          >
            Reserved
          </span>
          Description 1
        </p>
        <button
          class="dragon-btn cancel"
          data-testid="cancel-reservation-button-1"
          type="button"
        >
          Cancel Reservation
        </button>
      </div>
    </div>
    <div
      class="dragon-left"
    >
      <img
        alt="Dragon 2"
        class="dragon-image"
        src="dragon2.jpg"
      />
      <div
        class="dragon-details"
      >
        <h2
          class="dragon-name"
        >
          Dragon 2
        </h2>
        <p
          class="dragon-desc"
        >
          Description 2
        </p>
        <button
          class="dragon-btn"
          data-testid="reserve-button-2"
          type="button"
        >
          Reserve dragon
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
